File: ios/Runner/Casting/CastingPlaybackProvider.swift
import Foundation

struct ProviderState {
    let connectionState: CastingConnectionState
    let playbackState: CastingPlaybackState
    let device: CastDevice?
    let positionMs: Int64
    let durationMs: Int64
    let error: String?

    init(
        connectionState: CastingConnectionState = .disconnected,
        playbackState: CastingPlaybackState = .idle,
        device: CastDevice? = nil,
        positionMs: Int64 = 0,
        durationMs: Int64 = 0,
        error: String? = nil
    ) {
        self.connectionState = connectionState
        self.playbackState = playbackState
        self.device = device
        self.positionMs = positionMs
        self.durationMs = durationMs
        self.error = error
    }
}

@MainActor
protocol CastingPlaybackProvider: AnyObject {
    func loadMedia(_ mediaInfo: MediaInfo, autoplay: Bool, positionMs: Int64)
    func play()
    func pause()
    func seek(to positionMs: Int64)
    func stop()
    func setVolume(_ volume: Float)
    func disconnect()
}

File: android/app/src/main/kotlin/dev/rohanjsh/pitcher/casting/CastingPlaybackProvider.kt
package dev.rohanjsh.pitcher.casting

interface CastingPlaybackProvider {
    fun setStateChangeHandler(handler: StateChangeHandler)
    fun loadMedia(mediaInfo: MediaInfo, autoplay: Boolean, positionMs: Long)
    fun play()
    fun pause()
    fun seek(positionMs: Long)
    fun stop()
    fun setVolume(volume: Double)
    fun disconnect()

    fun interface StateChangeHandler {
        fun onStateChanged(state: ProviderState)
    }
}

data class ProviderState(
    val connectionState: CastingConnectionState = CastingConnectionState.DISCONNECTED,
    val playbackState: CastingPlaybackState = CastingPlaybackState.IDLE,
    val device: CastDevice? = null,
    val positionMs: Long = 0,
    val durationMs: Long = 0,
    val error: String? = null
)

